internal/integrity_check_runner.go

```diff
@@ -0,0 +1,298 @@
+package internal
+
+import (
+	"github.com/pkg/errors"
+	"github.com/wal-g/storages/storage"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/utility"
+	"sort"
+)
+
+// IntegrityCheckRunner queries the current cluster WAL segment and timeline
+// and travels through WAL segments in storage in reversed chronological order (starting from that segment)
+// to find any missing WAL segments that could potentially fail the PITR procedure
+type IntegrityCheckRunner struct {
+	startWalSegment           WalSegmentDescription
+	stopWalSegmentNo          WalSegmentNo
+	uploadingSegmentRangeSize int
+	storageFileNames          []string
+	timelineSwitchMap         map[WalSegmentNo]*TimelineHistoryRecord
+}
+
+func NewIntegrityCheckRunner(rootFolder storage.Folder, currentWalSegment WalSegmentDescription) (IntegrityCheckRunner, error) {
+	walFolder := rootFolder.GetSubFolder(utility.WalPath)
+
+	timelineSwitchMap, err := createTimelineSwitchMap(currentWalSegment.Timeline, walFolder)
+	if err != nil {
+		return IntegrityCheckRunner{}, errors.Wrap(err, "Failed to initialize timeline history map")
+	}
+
+	stopWalSegmentNo, err := getEarliestBackupStartSegmentNo(timelineSwitchMap, currentWalSegment.Timeline, rootFolder)
```

Comment:
I guess the logic is the same as was in wal_handler.go? And now you check the earliest backup also.

Reply:
yes